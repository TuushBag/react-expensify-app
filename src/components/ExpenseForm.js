import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            textarea: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        let description = e.target.value
        this.setState(() => ({ description }))
    }

    onTextAreaChange = (e) => {
        let textarea = e.target.value
        this.setState(() => ({ textarea })) 
    }
    
    onAmountChange = (e) => {
        let amount  = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }  
    }
    
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.description || !this.state.amount) {
            let errMessage = 'Please enter a description or an amount'
            this.setState(() => ({ error: errMessage }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.textarea
            })
        }
    }

render() {
     return (
     <div>
         { this.state.error && <p>{ this.state.error }</p>}
         <form onSubmit={this.onSubmit}>
              <input autoFocus
                 type='text' 
                 placeholder='description' 
                 value={this.state.description} 
                 onChange={this.onDescriptionChange}/>

               <input 
                type='number' 
                placeholder='amount' 
                value={this.state.amount} 
                onChange={this.onAmountChange}/>
    
            <SingleDatePicker 
              date={this.state.createdAt}
              onDateChange={this.onDateChange}
              focused={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />

            <textarea 
            placeholder='A descritption about your expense' 
            value={this.state.textarea} 
            onChange={this.onTextAreaChange}>
            </textarea>

            <button>Add expense</button>
        </form>          
    </div>  
        )
    }
}