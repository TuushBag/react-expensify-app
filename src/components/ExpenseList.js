import React from 'react'
import ExpenseListItem from './ExpenseListItem'
import selectedExpense from '../selectors/expenses'
import { connect } from 'react-redux'

 export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
               <h2>No expenses found</h2>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />
                })
            )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectedExpense( state.expenses, state.filters )
    }
} 

export default connect(mapStateToProps)(ExpenseList)

