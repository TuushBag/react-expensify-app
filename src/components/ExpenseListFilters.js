import React from 'react'
import { setTextFilter, sortByDate, sortByAmount, setByStartDate, setByEndDate} from '../actions/filters'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'


export class ExpenseListFilters extends React.Component {
    state = {
       calenderFocused: null
    }
     
    onDateChange = ({ startDate, endDate }) => {
       this.props.dispatch(setByStartDate(startDate))
       this.props.dispatch(setByEndDate(endDate))
    }

    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }))
    }

    onTextChange = (e) => {
        this.props.dispatch(setTextFilter(e.target.value))
    }

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.dispatch(sortByDate())
        } else if (e.target.value === 'amount') {
            this.props.dispatch(sortByAmount())
        }
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={this.onTextChange} />
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>

                <DateRangePicker 
                  startDate={this.props.filters.startDate}
                  endDate={this.props.filters.endDate}
                  onDatesChange={this.onDateChange}
                  focusedInput={this.state.calenderFocused}
                  onFocusChange={this.onFocusChange}
                  showClearDates={true}
                  numberOfMonths={1}
                  isOutsideRange={() => false}
                />
            </div>
        )  
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters,
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)