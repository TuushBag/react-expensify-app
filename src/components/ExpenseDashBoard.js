import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashBoard = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
 )

 export default ExpenseDashBoard