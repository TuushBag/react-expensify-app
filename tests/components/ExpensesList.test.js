import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../src/components/ExpenseList'
import expenses from '../fixtures/expenseData'

test('should test the snapshot of the expense filter', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should test when the expense array is empty', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})
