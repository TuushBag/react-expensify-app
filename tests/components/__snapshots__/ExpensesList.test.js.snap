// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the snapshot of the expense filter 1`] = `
<div>
  <ExpenseListItem
    amount={1293}
    createdAt={0}
    description="Internet"
    id="1"
    key="1"
    note="Harness the power of the internet"
  />
  <ExpenseListItem
    amount={293}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note="Settling your rent"
  />
  <ExpenseListItem
    amount={123}
    createdAt={345600000}
    description="Light"
    id="3"
    key="3"
    note="Electricity bill"
  />
</div>
`;

exports[`should test when the expense array is empty 1`] = `
<div>
  <h2>
    No expenses found
  </h2>
</div>
`;
